api/schedule/create.go

```diff
@@ -0,0 +1,230 @@
+//     schema:
+//       "$ref": "#/definitions/Error"
+//   '503':
+//     description: Unable to create the schedule
+//     schema:
+//       "$ref": "#/definitions/Error"
+
+// CreateSchedule represents the API handler to
+// create a schedule in the configured backend.
+func CreateSchedule(c *gin.Context) {
+	// capture middleware values
+	u := user.Retrieve(c)
+	r := repo.Retrieve(c)
+	allowlist := c.Value("allowlistschedule").([]string)
+	minimumFrequency := c.Value("scheduleminimumfrequency").(time.Duration)
+
+	// capture body from API request
+	input := new(library.Schedule)
+
+	err := c.Bind(input)
+	if err != nil {
+		retErr := fmt.Errorf("unable to decode JSON for new schedule: %w", err)
+
+		util.HandleError(c, http.StatusBadRequest, retErr)
+
+		return
+	}
+
+	// ensure the entry is valid
+	err = validateEntry(minimumFrequency, input.GetEntry())
+	if err != nil {
+		retErr := fmt.Errorf("schedule of %s is invalid: %w", input.GetName(), err)
+
+		util.HandleError(c, http.StatusBadRequest, retErr)
+
+		return
+	}
+
+	// ensure schedule name is defined
+	if input.GetName() == "" {
+		util.HandleError(c, http.StatusBadRequest, fmt.Errorf("schedule name must be set"))
+	}
+
+	// update engine logger with API metadata
+	//
+	// https://pkg.go.dev/github.com/sirupsen/logrus?tab=doc#Entry.WithFields
+	logrus.WithFields(logrus.Fields{
+		"org":  r.GetOrg(),
+		"repo": r.GetName(),
+		"user": u.GetName(),
+	}).Infof("creating new schedule %s", input.GetName())
+
+	// ensure repo is allowed to be activated
+	if !util.CheckAllowlist(r, allowlist) {
+		retErr := fmt.Errorf("unable to create schedule %s: %s is not on allowlist", input.GetName(), r.GetFullName())
+
+		util.HandleError(c, http.StatusForbidden, retErr)
+
+		return
+	}
+
+	s := new(library.Schedule)
+
+	// update fields in schedule object
+	s.SetCreatedBy(u.GetName())
+	s.SetRepoID(r.GetID())
+	s.SetName(input.GetName())
+	s.SetEntry(input.GetEntry())
+	s.SetCreatedAt(time.Now().UTC().Unix())
+	s.SetUpdatedAt(time.Now().UTC().Unix())
+	s.SetUpdatedBy(u.GetName())
+
+	// set the active field based off the input provided
+	if input.Active == nil {
+		// default active field to true
+		s.SetActive(true)
+	} else {
+		s.SetActive(input.GetActive())
+	}
+
+	// send API call to capture the schedule from the database
+	dbSchedule, err := database.FromContext(c).GetScheduleForRepo(r, input.GetName())
+	if err == nil && dbSchedule.GetActive() {
+		retErr := fmt.Errorf("unable to create schedule: %s is already active", input.GetName())
+
+		util.HandleError(c, http.StatusConflict, retErr)
+
+		return
+	}
+
+	if !r.GetActive() {
+		retErr := fmt.Errorf("unable to create schedule: %s repo %s is disabled", input.GetName(), r.GetFullName())
+
+		util.HandleError(c, http.StatusConflict, retErr)
+
+		return
+	}
+
+	// if the repo exists but is inactive
+	if len(r.GetOrg()) > 0 && !dbSchedule.GetActive() && input.GetActive() {
+		// update the repo owner
```

Comment:
Thoughts on updating the wording here slightly?

```suggestion
		// update the user who created the schedule
```